File: leetcode/c++/199_binary_tree_right_side_view.cpp
// https://leetcode.cn/problems/binary-tree-right-side-view

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    void dfs(TreeNode* root, vector<int>& res, int height) {
        if (root == nullptr)
            return;
        if (height == res.size()) 
            res.push_back(root->val);
        dfs(root->right, res, height+1);
        dfs(root->left, res, height+1);
    }
    vector<int> rightSideView(TreeNode* root) {
        vector<int>res;
        dfs(root, res, 0);
        return res;
    }   
};

// bfs和dfs都能做出来，bfs就是每一层最后一个添加到数组中；dfs就是根节点，右子节点，左子节点这样递归，第一次到这个高度时就添加进数组
// 时间复杂度和空间复杂度两者都是O(n)

File: leetcode/golang/199_binary_tree_right_side_view.go
type TreeNode struct {
    Val int
    Left *TreeNode
    Right *TreeNode
}

func rightSideView(root *TreeNode) []int {
    var que []*TreeNode
    var res []int
    if root != nil {
        que = append(que, root)
    }
    for len(que) != 0 {
        for i := len(que)-1; i >= 0; i-- {
            node := que[0]
            que = que[1:len(que)]
            if i == 0 {
                res = append(res, node.Val)
            }
            if node.Left != nil {
                que = append(que, node.Left)
            }
            if node.Right != nil {
                que = append(que, node.Right)
            }
        }
        
    }
    return res
}